### ISSUE TITLE
storage: SSTable file leak

### ISSUE THREAD

**Issue by mrtracy** [cockroachdb/cockroach#23117]
Due to an unknown cause, SSTable files are leaking on the latest 2.0 alphas. This can be seen easily in long running clusters, where the count of .sst files on the file system is significantly higher than the number of SSTables.  For example, on one of our test clusters:

![screen shot 2018-02-26 at 4 42 24 pm](https://user-images.githubusercontent.com/6969858/36697140-2e22c09e-1b14-11e8-9f98-03456135b19b.png)

From SSH:
```
>:/mnt/data1# ls -1 *.sst | wc -l
826
```

Note that the number of open file descriptors seems to be more in line with the number of files:
![screen shot 2018-02-26 at 4 42 36 pm](https://user-images.githubusercontent.com/6969858/36697167-475922ba-1b14-11e8-87f6-22816fffe2bf.png)

This can cause issues because these leaks can cause disk usage to rapidly spike on clusters under certain activity.  An example of this (which caused the node in question to crash):

![pasted image at 2018_02_26 03_35 pm](https://user-images.githubusercontent.com/6969858/36697225-760ce0ec-1b14-11e8-8f2b-6079119dd789.png)

Note that the node in question had just come back online after being down; this might suggest that leak might be caused by activity related to snapshots (although this is based on circumstantial evidence).

**Comment by bdarnell** [comment]
Looks like something changed on adriatic around 2/19:

![image](https://user-images.githubusercontent.com/160562/36697987-a1417dd4-1b16-11e8-8811-06a2ec3d1ba8.png)

**Comment by bdarnell** [comment]
Looking at adriatic-5, `ls` and `lsof` are in agreement about the number of sstable files, so all the files are open simultaneously:

```
root@cockroach-adriatic-0005:/mnt/data1# lsof -p 7514|grep '\.sst'|wc
   1212   10908  100596
root@cockroach-adriatic-0005:/mnt/data1# ls -l *.sst|wc
   1211   10899   71449
```

Rocksdb's live sstable count is 659 at the same time. This is consistent with the theory that the root cause is some sort of iterator/snapshot leak holding these files open; they're not getting completely lost to the process.

**Comment by benesch** [comment]
@bdarnell where did you pull those commit SHAs from? We skipped a number of releases; it seems like this is the range:

![screenshot 2018-02-26 18 50 00](https://user-images.githubusercontent.com/882976/36702635-e2f85dce-1b25-11e8-943a-3c2436dcac39.png)

https://github.com/cockroachdb/cockroach/compare/886f1f3...c02d5061f8623b716a14016ea4a90d0dbb245691

**Comment by bdarnell** [comment]
From `ls -l` on the machine. 

```
root@cockroach-adriatic-0005:~# ls -lrt cockroach*
-rwxr-xr-x 1 root root 54488952 Jan 24 19:06 cockroach.linux-gnu-amd64.4a55988875439cde88c67bc70e30e8d1d181c7f3
-rwxr-xr-x 1 root root 54508056 Jan 24 23:11 cockroach.linux-gnu-amd64.624cd615a99c86348f612d0a824759a990fae197
-rwxr-xr-x 1 root root 46287912 Jan 29 21:49 cockroach.linux-gnu-amd64.ec52ded95a5235d71f26b6cbc59924bd6733fb3a
-rwxr-xr-x 1 1002 1003 46287912 Feb  5 17:41 cockroach.v1.1.5
-rwxr-xr-x 1 root root 54712696 Feb  5 18:18 cockroach.linux-gnu-amd64.886f1f34c0c07dc1e9ee205ae6a90d4a80a611f9
-rwxr-xr-x 1 root root 55543064 Feb 13 20:04 cockroach.linux-gnu-amd64.f27753540d4ef6469d446e83493b7b5645ea9eef
-rwxr-xr-x 1 root root 55590040 Feb 13 22:56 cockroach.linux-gnu-amd64.c02d5061f8623b716a14016ea4a90d0dbb245691
-rwxr-xr-x 1 root root 56145304 Feb 20 00:33 cockroach.linux-gnu-amd64.466c1853e75ac5a23a6b4fdeb6a9a3aaf66743ab
-rwxr-xr-x 1 root root 56193976 Feb 22 03:50 cockroach.linux-gnu-amd64.efa7583c93b933c5321cda7fe08a11b22e925985
-rwxr-xr-x 1 root root 56222840 Feb 24 04:32 cockroach.linux-gnu-amd64.ae07c0ba6184a8e40a6369371cbc8618a9ef1477
lrwxrwxrwx 1 root root       66 Feb 24 04:32 cockroach -> cockroach.linux-gnu-amd64.ae07c0ba6184a8e40a6369371cbc8618a9ef1477
```

**Comment by benesch** [comment]
<img width="850" alt="screenshot 2018-02-26 22 25 12" src="https://user-images.githubusercontent.com/882976/36709050-ecb3675a-1b43-11e8-8ca4-eed45a581098.png">

The sawtooth we were seeing when we brought node 3 back online has vanished. This gives weight to @spencerkimball's theory that we had a backlog of compaction activity from when the node lost its leases.

Still, given that adriatic is running with a smallish cycle length, it should never have run out of disk space in the first place. Nodes 2 and 3 have been running with my time-bomb iterators (heh) for the past few hours and haven't exploded yet, so it looks like it's more subtle than a leaked iterator.

**Comment by benesch** [comment]
The more I look at it, the more I think the leaked iterator in migrateLegacyTombstones is to blame (#23136). Adriatic 4-6 are still running the alpha candidate and each have a few hundred leaked SSTs. Adriatic 1-3, by contrast, have binaries with #23136 applied and RocksDB's SST count exactly matches  the on-disk count, except during compactions.

Here's adriatic 1 in the ten minutes after I deployed #23136;

<img width="848" alt="screenshot 2018-02-26 23 08 56" src="https://user-images.githubusercontent.com/882976/36710135-1392c1e4-1b4a-11e8-9f7e-ebf8b06f4bc5.png">

Interesting, isn't it? I wonder if keeping all those SSTs open was preventing compactions from being as effective as normal.

**Comment by bdarnell** [comment]
Yeah, #23136 is probably enough to explain this. A single leaked iterator is enough to eventually double the data size in this high-churn workload, and maybe worse if it also causes log retention.

The legacy tombstone migration is sneaky - it runs on each restart after the cluster version setting has been bumped, but we don't normally update the cluster setting when upgrading adriatic. We did run it on 2/13, though, which set this migration up to run on the next restart:

    I180213 22:59:43.195720 88928 sql/event_log.go:114  [client=127.0.0.1:38056,user=root,n1] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:(SELECT crdb_internal.node_executable_version()) User:root}

**Comment by a-robinson** [comment]
Fixed by #23136/#23153, right @benesch?

**Comment by benesch** [comment]
Yep, was just verifying! Adriatic's SST count exactly matches the on-disk count on all six of its nodes since 59246c8 was deployed. Considering this fixed.
